### PR TITLE
Add diff

### PR DESCRIPTION
**Pull request by msluszniak**

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -10987,6 +10987,69 @@ defmodule Nx do
...
+
+      iex> Nx.diff(Nx.tensor([[1, 3, 6, 10], [0, 5, 6, 8]]), axis: 0)
+      #Nx.Tensor<
+        s64[1][4]
+        [
+          [-1, 2, 0, -2]
+        ]
+      >
+  """
+  @doc type: :cumulative
```
Comment (msluszniak) [inline nx/lib/nx.ex:11032]: I have no idea if this is the right category

Comment (polvalente) [reply nx/lib/nx.ex:11041]: Given what we have, maybe window? Give it can be implemented as a convolution as well?

Comment (msluszniak) [reply nx/lib/nx.ex:11041]: Hmm, the 2D analogy isn't much intuitive for me.

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -10987,6 +10987,86 @@ defmodule Nx do
...
+  def diff(tensor, opts \\ []) do
+    opts = keyword!(opts, n: 1, axis: -1)
+    %T{shape: shape, names: names} = tensor = to_tensor(tensor)
+    n = opts[:n]
+    axis = Nx.Shape.normalize_axis(shape, opts[:axis], names)
+
+    if rank(tensor) == 0 do
+      raise ArgumentError, "Cannot compute diff of a scalar"
```
Comment (josevalim) [inline nx/lib/nx.ex:11049]:
```suggestion
      raise ArgumentError, "cannot compute diff of a scalar"
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -10987,6 +10987,86 @@ defmodule Nx do
...
+    axis_size = Nx.axis_size(tensor, axis)
+
+    if n == 0 do
+      tensor
+    else
+      Enum.reduce(0..(n - 1), tensor, fn x, acc ->
+        subtract(
+          take(acc, add(1, iota({axis_size - x - 1})), axis: axis),
+          take(acc, iota({axis_size - x - 1}), axis: axis)
+        )
```
Comment (polvalente) [inline nx/lib/nx.ex:11065]:
```suggestion
        subtract(
          slice_along_axis(acc, 1, axis_size - x - 1, axis: axis),
          slice_along_axis(acc, 0, axis_size - x - 1, axis: axis)
        )
```

Comment (msluszniak) [reply nx/lib/nx.ex:11065]: It should be slice_along_axis(acc, 1, axis_size - x, axis: axis) in the first one, but yes

**Review by polvalente — APPROVED** [review]

**Pull request merged**